docs/explanation/locks.rst

```diff
@@ -0,0 +1,236 @@
+On `SMP systems`_ (Symmetric Multiprocessing), the Linux kernel provides various synchronization mechanisms to safeguard shared data structures from concurrent access, preventing `race conditions`_ in the kernel space.
+
+Locks are primarily used to prevent race conditions, which are a concern in both SMP and real-time systems.
+Even in a single-processor real-time system, the preemption mechanism can lead to race conditions similar to those found in an SMP system.
+
+There are essentially three classes of locks in the kernel:
+
+- `Spinning locks`_
+- `Sleeping locks`_ (also known as blocking locks)
+- `CPU local locks`_
+
+Spinlocks
+----------
+
+Spinlocks (found in `include/linux/spinlock.h`_) are locks that keep polling (or "spinning"), attempting to acquire the lock.
+They are small, fast, and simple to use, making them generally applicable throughout the kernel.
+Spinlocks are ideal for situations where the kernel process cannot sleep.
+The ideal type of spinlock to use depends on the specific contexts between which the data needs to be shared.
+
+Within the class of spinlocks, there are various types:
+
+.. _rw-lock:
+
+rwlock_t
+~~~~~~~~
+
+``rwlock_t`` (defined on `include/linux/rwlock_types.h`_)  is a variant of the read/write lock for spinning locks.
+Consumers of this lock are divided into two classes: readers and writers.
+This lock structure allows multiple readers but only one writer at a time.
+By implementing this approach, the number of locks used can be reduced, which in turn decreases the overhead associated with locking.
+
+.. _spinlock:
+
+spinlock_t
+~~~~~~~~~~
+
+``spinlock_t`` is the most generic implementation of spinlock structures, as defined in `include/linux/spinlock_types.h`_.
+It is implemented in two different ways, depending on whether the real-time kernel configuration (i.e., with ``PREEMPT_RT`` enabled) is used.
+
+- In non-PREEMPT_RT scenarios, ``spinlock_t`` is a ``typedef`` structure based on :ref:`raw_spinlock`.
+- In preemptive scenarios (real-time kernels), ``spinlock_t`` is implemented as an :ref:`rt_mutex`, meaning it uses an ``rt_mutex`` in its ``typedef`` structure. This means that ``PREEMPT_RT`` changes the lock category of ``spinlock_t`` from a spinning lock to a sleeping lock.
+
+.. _raw_spinlock:
+
+raw_spinlock_t
+~~~~~~~~~~~~~~
+
+``raw_spinlock_t`` (defined on `include/linux/spinlock_types_raw.h`_) is a strict spinning lock implementation that operates consistently across all kernels, including real-time kernels.
+It is reserved for critical core code, low-level interrupt handling, and scenarios where disabling preemption or interrupts is necessary.
+For example, this lock is used to safely access hardware state.
+In some cases, ``raw_spinlock_t`` can also be employed when the critical section is very small, thereby avoiding the overhead associated with :ref:`rt_mutex`.
+
+Spinlock operations
+~~~~~~~~~~~~~~~~~~~
```

Comment:
This heading is at the same level with spinlock types. Move all spinlock types to one level lower under its own subsection?

Reply:
I'm wondering if these are in the scope of this document. Good explanations there, so maybe they should be moved to a separate document.

Reply:
I've changed Spinlock Operations to H2 and all the operations to H3.

Reply:
Please remove from this version of the doc. They can be added together with the documentation of other operations.

Reply:
Diff for removal: https://github.com/canonical/real-time-ubuntu-docs/compare/f205cd1d5499bf93b7da6c7e5b08a1c577cd6c34..6d2bd2e09a6011394d14c247daa226ca72b92727